InterfacePython/example_sparse_grids_05.py

```diff
@@ -0,0 +1,121 @@
+
+import matplotlib.pyplot as pyplot
+
+def example_05():
+
+    print("\n---------------------------------------------------------------------------------------------------\n")
+    print("Example 5: interpolate f(x,y) = exp(-x^2) * cos(y), using leja rule")
+    print("           employ adaptive refinement to increase accuracy per samples")
+
+    iNumInputs = 2
+    iNumOutputs = 1
+    def model(aX):
+        print(aX, np.ones((1,)) * np.exp(-aX[0] * aX[0]) * np.cos(aX[1]))
+        return np.ones((1,)) * np.exp(-aX[0] * aX[0]) * np.cos(aX[1])
+
+    iTestGridSize = 5
+    dx = np.linspace(-1.0, 1.0, iTestGridSize) # sample on a uniform grid
+    aMeshX, aMeshY = np.meshgrid(dx, dx)
+    aTestPoints = np.column_stack([aMeshX.reshape((iTestGridSize**2, 1)),
+                                   aMeshY.reshape((iTestGridSize**2, 1))])
+    aReferenceValues = np.exp(-aTestPoints[:,0]**2) * np.cos(aTestPoints[:,1])
+
+    def testGrid(grid, aTestPoints, aReferenceValues):
+        aResult = grid.evaluateBatch(aTestPoints)
+        for i in range(20):
+            aX = aTestPoints[i,:]
+        #    print(model(aX), aResult[i], aReferenceValues[i])
+        #print(np.abs(aResult - aReferenceValues))
+        #exit(1)
+        return np.max(np.abs(aResult - aReferenceValues))
+
+
+    iInitialLevel = 6
+
+    grid_isotropic = Tasmanian.SparseGrid()
+    grid_iptotal   = Tasmanian.SparseGrid()
+    grid_icurved   = Tasmanian.SparseGrid()
+    grid_surplus   = Tasmanian.SparseGrid()
+    grid_isotropic.makeGlobalGrid(iNumInputs, iNumOutputs, iInitialLevel, "level", "leja")
+    grid_iptotal.copyGrid(grid_isotropic)
+    grid_icurved.copyGrid(grid_isotropic)
+    grid_surplus.copyGrid(grid_isotropic)
+
+    iNumThreads = 1
+    iBudget = 100
+    print("{0:>22s}{1:>22s}{2:>22s}{3:>22s}".format("isotropic", "iptotal", "ipcurved", "surplus"))
+    print("{0:>8s}{1:>14s}{0:>8s}{1:>14s}{0:>8s}{1:>14s}{0:>8s}{1:>14s}".format("points", "error"))
+
+    bBelowBudget = True
+    while(bBelowBudget):
+        sInfo = ""
+        if (grid_isotropic.getNumLoaded() < iBudget):
+            Tasmanian.loadNeededPoints(lambda x, tid : model(x), grid_isotropic, iNumThreads)
+            sInfo += "{0:>8d}{1:>14s}".format(grid_isotropic.getNumLoaded(),
+                     "{0:1.4e}".format(testGrid(grid_isotropic, aTestPoints, aReferenceValues)))
+
+            iLevel = 0
+            while(grid_isotropic.getNumNeeded() == 0):
+                grid_isotropic.updateGlobalGrid(iLevel, "level")
+                iLevel += 1
+        else:
+            sInfo += "{0:>14s}".format("")
+
+
+
+        sInfo += "{0:>14s}".format("")
+        print(sInfo)
+        bBelowBudget = (grid_isotropic.getNumLoaded() < iBudget
+                       or grid_icurved.getNumLoaded() < iBudget
+                       or grid_icurved.getNumLoaded() < iBudget
+                       or grid_surplus.getNumLoaded() < iBudget)
+
+        bBelowBudget = (grid_isotropic.getNumLoaded() < iBudget)
```

Comment:
Did you mean this 2nd assignment?

Reply:
I am still debugging this, without the second assignment it will cause an infinite loop. It won't  when I add all other grids.